Extension/src/ghAPI.ts

```diff
@@ -0,0 +1,152 @@
+                switch (platformInfo.architecture) {
+                    case 'x86': return 'cpptools-linux32.vsix';
+                    case 'x86_64': return 'cpptools-linux.vsix';
+                }
+            }
+        }
+    }(platformInfo);
+    if (!vsixName) {
+        return;
+    }
+
+    // Get the URL to download the VSIX, using vsixName as a key
+    const downloadUrl: string = build.assets.find((asset) => {
+        return asset.name === vsixName;
+    }).browserDownloadUrl;
+
+    return downloadUrl;
+}
+
+export async function getTargetBuildURL(updateChannel: string): Promise<string> {
+    const releaseJson: Build[] = await getReleaseJson();
+    if (!releaseJson) {
+        return;
+    }
+
+    const targetRelease: Build = getTargetBuild(releaseJson, updateChannel);
+    if (!targetRelease) {
+        return;
+    }
+
+    const downloadUrl: string = await downloadUrlForPlatform(targetRelease);
+    if (!downloadUrl) {
+        return;
+    }
+}
+
+// Determines whether there exists a build that should be installed; returns the build if there is
+function getTargetBuild(releaseJson: Build[], updateChannel: string): Build {
+    // Get predicates to determine the build to install, if any
+    let needsUpdate: (v1: PackageVersion, v2: PackageVersion) => boolean;
+    let useBuild: (build: Build) => boolean;
+    if (updateChannel === 'Insiders') {
+        needsUpdate = function(v1: PackageVersion, v2: PackageVersion): boolean { return v1.isGreaterThan(v2); };
+        useBuild = function(build: Build): boolean { return true; };
+    } else if (updateChannel === 'Default') {
+        needsUpdate = function(v1: PackageVersion, v2: PackageVersion): boolean { return v2.isGreaterThan(v1); };
+        useBuild = function(build: Build): boolean { return build.name.indexOf('-') === -1; };
+    } else {
+        return;
+    }
+
+    // Get the build to install
+    const targetBuild: Build = releaseJson.find((build) => useBuild(build));
+    if (!targetBuild) {
+        return;
+    }
+
+    // Check current version against target's version to determine if the installation should happen
+    const userVersion: PackageVersion = new PackageVersion(util.packageJson.version);
+    const targetVersion: PackageVersion = new PackageVersion(targetBuild.name);
+    if (!userVersion.isValid || !targetVersion.isValid) {
+        return;
+    }
+    if (!needsUpdate(userVersion, targetVersion)) {
+        return;
+    }
+
+    return targetBuild;
+}
+
+async function getReleaseJson(): Promise<Build[]> {
+    const releaseJsonFile: any = tmp.fileSync();
+
+    // Download json from GitHub
+    const releaseUrl: string = 'https://api.github.com/repos/Microsoft/vscode-cpptools/releases';
+    await util.downloadFileToDestination(releaseUrl, releaseJsonFile.name, { 'User-Agent': 'vscode-cpptools' }).catch(() => {
+            // TODO check internet connection before attempting? That way we can mitigate false positive telemetry
```

Comment:
It's hard to determine what to check.  They could have Internet, but be behind a proxy and they haven't configured it correctly.  It's good for us to see those errors to we can attempt to mitigate them.  You can remove this comment